plugins/paytovote/paytovote.go

```diff
@@ -0,0 +1,183 @@
+func New(name string) *P2VPlugin {
+	return &P2VPlugin{
+		name: name,
+	}
+}
+
+func newState(issue string) P2VPluginState {
+	return P2VPluginState{
+		TotalCost:    types.Coins{},
+		Issue:        issue,
+		votesFor:     0,
+		votesAgainst: 0,
+		votesSpoiled: 0,
+	}
+}
+
+func (p2v *P2VPlugin) SetOption(store types.KVStore, key string, value string) (log string) {
+	return ""
+}
+
+func (p2v *P2VPlugin) RunTx(store types.KVStore, ctx types.CallContext, txBytes []byte) (res abci.Result) {
+
+	// Decode tx
+	var tx P2VTx
+	err := wire.ReadBinaryBytes(txBytes, &tx)
+	if err != nil {
+		return abci.ErrBaseEncodingError.AppendLog("Error decoding tx: " + err.Error())
+	}
+
+	// Validate tx
+	if !tx.Valid {
+		return abci.ErrInternalError.AppendLog("P2VTx.Valid must be true")
+	}
+
+	if len(tx.Issue) == 0 {
+		return abci.ErrInternalError.AppendLog("P2VTx.Issue must have a length greater than 0")
+	}
+
+	if !tx.Cost2Vote.IsValid() {
+		return abci.ErrInternalError.AppendLog("P2VTx.Cost2Vote is not sorted or has zero amounts")
+	}
+
+	if !tx.Cost2Vote.IsNonnegative() {
+		return abci.ErrInternalError.AppendLog("P2VTx.Cost2Vote must be nonnegative")
+	}
+
+	if !tx.Cost2CreateIssue.IsValid() {
+		return abci.ErrInternalError.AppendLog("P2VTx.Cost2CreateIssue is not sorted or has zero amounts")
+	}
+
+	if !tx.Cost2CreateIssue.IsNonnegative() {
+		return abci.ErrInternalError.AppendLog("P2VTx.Cost2CreateIssue must be nonnegative")
+	}
+
+	// Load P2VPluginState
+	var p2vState P2VPluginState
+	p2vStateBytes := store.Get(p2v.StateKey(tx.Issue))
+
+	//Determine if the issue already exists
+	issueExists := true
+
+	if len(p2vStateBytes) > 0 { //is there a record of the issue existing?
+		err = wire.ReadBinaryBytes(p2vStateBytes, &p2vState)
+		if err != nil {
+			return abci.ErrInternalError.AppendLog("Error decoding state: " + err.Error())
+		}
+	} else {
+		issueExists = false
+	}
+
+	returnLeftover := func(cost types.Coins) {
+		leftoverCoins := ctx.Coins.Minus(cost)
+		if !leftoverCoins.IsZero() {
+			// TODO If there are any funds left over, return funds.
+			// ctx.CallerAccount is synced w/ store, so just modify that and store it.
+		}
+	}
+
+	switch {
+	case tx.ActionTypeByte == TypeByteCreateIssue && issueExists:
+		return abci.ErrInsufficientFunds.AppendLog("Cannot create an already existing issue")
+	case tx.ActionTypeByte != TypeByteCreateIssue && !issueExists:
+		return abci.ErrInsufficientFunds.AppendLog("Tx Issue not found")
+	case tx.ActionTypeByte == TypeByteCreateIssue && !issueExists:
+		// Did the caller provide enough coins?
+		if !ctx.Coins.IsGTE(tx.Cost2CreateIssue) {
+			return abci.ErrInsufficientFunds.AppendLog("Tx Funds insufficient for creating a new issue")
+		}
+
+		// Update P2VPluginState
+		newP2VState := newState(tx.Issue)
+		newP2VState.TotalCost = newP2VState.TotalCost.Plus(tx.Cost2Vote)
+
+		// Save P2VPluginState
+		store.Set(p2v.StateKey(tx.Issue), wire.BinaryBytes(newP2VState))
+
+		returnLeftover(tx.Cost2CreateIssue)
```

Comment:
Yes, I have this setup to return within the returnLeftover function variable (which needs to be completed, this is a remaining issue) Maybe renaming returnLeftover to something less ambiguous that implies that it's subtracting cost resolves this?https://github.com/tendermint/basecoin/blob/bfac5437c34dc513757c56c048531f04ea3acf60/plugins/paytovote/paytovote.go#L119